src/combinations.rs

```diff
@@ -27,22 +27,70 @@ impl<I> fmt::Debug for Combinations<I>
     debug_fmt_fields!(Combinations, indices, pool, first);
 }
 
-/// Create a new `Combinations` from a clonable iterator.
+/// Create a new `Combinations` from a cloneable iterator.
 pub fn combinations<I>(iter: I, k: usize) -> Combinations<I>
     where I: Iterator
 {
-    let mut pool: LazyBuffer<I> = LazyBuffer::new(iter);
+    let mut pool = LazyBuffer::new(iter);
+    pool.prefill(k);
 
-    for _ in 0..k {
-        if !pool.get_next() {
-            break;
+    Combinations::from_pool(pool, k)
+}
+
+impl<I: Iterator> Combinations<I> {
+    #[inline]
+    pub(crate) fn from_pool(pool: LazyBuffer<I>, k: usize) -> Self {
+        Combinations {
+            indices: (0..k).collect(),
+            pool,
+            first: true,
         }
     }
 
-    Combinations {
-        indices: (0..k).collect(),
-        pool,
-        first: true,
+    /// Returns the length of a combination produced by this iterator.
+    #[inline]
+    pub fn k(&self) -> usize { self.indices.len() }
+
+    /// Returns the (current) length of the pool from which combination elements are
+    /// selected. This value can change between invocations of `next()` and `init()`.
+    #[inline]
+    pub(crate) fn n(&self) -> usize { self.pool.len() }
+
+    /// Returns a reference to the source iterator.
+    #[inline]
+    pub(crate) fn src(&self) -> &I { &self.pool.it }
+
+    /// Resets this `Combinations` back to an initial state for combinations of length
+    /// `k` over the same pool data source. If `k` is larger than the previous value
+    /// an attempt is made to prefill the pool so that it holds `k` elements.
+    ///
+    /// Returns `true` if the underlying pool is large enough to provide combinations
+    /// of length `k`, otherwise `false`.
+    pub(crate) fn reset(&mut self, k: usize) -> bool {
+        let prev_k = self.indices.len();
+        let mut filled = true;
+
+        // Shrink/grow indices vector, filling in new values if required.
+        let reset_limit: usize;
+        if k < prev_k {
+            self.indices.truncate(k);
+            reset_limit = k;
+        } else {
+            filled = self.pool.prefill(k);
+
+            for i in prev_k..k {
+                self.indices.push(i);
+            }
+            reset_limit = prev_k;
+        }
+
+        // Reset old indices.
+        for i in 0..reset_limit {
+            self.indices[i] = i;
+        }
+
+        self.first = true;
+        filled
```

Comment:
Nitpick: I think it would be nicer if we could go with fewer variables here:

Something like:
* `self.first=true // always done on reset`
* if `k<indices.len`, truncate and set items to `0,1,...,k`, return `true`
* otherwise `pool.prefill`, `indices.extend(indices.len..k)`, set first items to `0,1,...,k`, return `false`

This way, we would avoid the back-and-forth wrt `reset_limit`, and avoid the `mut filled`. As said, this seems more a stylistic thing.

Reply:
Agreed, I was attempting to avoid duplicating the 'reset loop' but your suggestion makes it a lot cleaner. I might be able to remove the need to return `bool` depending on the details of other changes.